### DISCUSSION TITLE
MasonInstallAll is missing

### DISCUSSION THREAD

**Discussion by adorostkar** [mason-org/mason.nvim#1618] [Q&A]
Hello all,
I have installed Mason through Lazy.nvim but when I try to use the command `MasonInstallAll` it is not there. Is there something I am missing?

here is my Lazy setup. Please tell me if something is wrong with this as I am fairly new with neovim
```
    {
        "williamboman/mason.nvim",
        opts = {
            ensure_installed = {
                "clangd",
                "lua-lsp",
                "bash-language-server",
                "clang-format",
            }
        },
        config = function()
            require('mason').setup()
        end,
    },
    {
        "williamboman/mason-lspconfig.nvim",
        dependencies = {
            "williamboman/mason.nvim",
        },
        config = function()
            require("mason").setup()
            require("mason-lspconfig").setup()

            require("mason-lspconfig").setup_handlers {
                -- The first entry (without a key) will be the default handler
                -- and will be called for each installed server that doesn't have
                -- a dedicated handler.
                function (server_name) -- default handler (optional)
                    require("lspconfig")[server_name].setup {}
                end,
                -- Next, you can provide a dedicated handler for specific servers.
                -- For example, a handler override for the `rust_analyzer`:
                -- ["rust_analyzer"] = function ()
                --     require("rust-tools").setup {}
                -- end
            }
        end,
    },
```

**Comment by WhoIsSethDaniel** [comment]
There doesn't seem to be any such command. https://github.com/williamboman/mason.nvim?tab=readme-ov-file#commands

**Reply by adorostkar** [reply]
Yes I see now. I wonder though in some guides it says to use `ensure_installed` with combination of `MasonInstallAll` to automatically install some packages. How does this work with Mason then?

**Reply by WhoIsSethDaniel** [reply]
Can you provide a link to these guides?

**Reply by adorostkar** [reply]
Sure this is the link 
https://www.youtube.com/watch?v=lsFoZIg-oDs&t=132s

However after digging into NVChad I figured that NVChad adds the command "MasonInstallAll" and looks for ensure_installed. So it is not mason specific command. Rather it is added through NVChad

**Reply by Viraj-S45** [reply]
Hi there,

I'm having the same problem as you. I'm following the same video tutorial and updating the plugins.lua file as instructed.  After saving the file, I'm running :MasonInstallAll, but it's not installing clangd and clang-format. However, lua-language-server does get installed, even though it's not mentioned in the plugins.lua file.

Here's my plugins.lua code, which looks exactly like what's shown in the video guide:
```
{
  "williamboman/mason.nvim",
  opts = {
    ensure_installed = {
      "clangd",
      "clang-format"
    }
  }
}
```
I've double-checked that Mason is installed in my Neovim environment, and I've also tried restarting Neovim after making changes to plugins.lua and running :MasonInstallAll again. Unfortunately, clangd and clang-format are still not getting installed.

**Reply by h3rshh** [reply]
> I've double-checked that Mason is installed in my Neovim environment, and I've also tried restarting Neovim after making changes to plugins.lua and running :MasonInstallAll again. Unfortunately, clangd and clang-format are still not getting installed.


Did you find any solutions ?

**Reply by advl** [reply]
It would only make sense to have it, since it would allow for headless nvim installs `nvim --headless +MasonInstallAll +qall` that would automatically read from the `ensure_installed` value

EDIT :

Looking at : 
- [mason.nvim/blob/main/lua/mason/api/command.lua ](https://github.com/williamboman/mason.nvim/blob/main/lua/mason/api/command.lua))
- [mason-lspconfig.nvim/blob/main/lua/mason-lspconfig/ensure_installed.lua](https://github.com/williamboman/mason-lspconfig.nvim/blob/main/lua/mason-lspconfig/ensure_installed.lua) 
We can write something like

```lua
-- MasonInstallAll command implementation
vim.api.nvim_create_user_command("MasonInstallAll", function()
    local lspconfig_to_package = require("mason-lspconfig.mappings.server").lspconfig_to_package
    local packages = {
        'bashls', 'cssls', 'graphql', 'lua_ls', 'rust_analyzer', 'somesass_ls',
        'stylelint_lsp', 'shellcheck', 'tsserver', 'yamlls',
    }

    local package_args = {}
    
    for _, package in ipairs(packages) do
        local server_name = lspconfig_to_package[package]
        if server_name then
            table.insert(package_args, server_name)
        else
            table.insert(package_args, package)
        end
    end                                                        
```
It does work headlessly. `nvim --headless "+MasonInstallAll" +qall`
You will probably want to modify this if youre installing only lsps, or only regular packages

**Reply by ryanpeach** [reply]
@advl This gives me `Not an editor command: MasonInstallAll`

**Reply by ryanpeach** [reply]
Also, several plugins in a config can add things to ensure_installed when their "config" function is run, how can we be sure they have all added their stuff if they are lazy loaded?

**Reply by paaloeye** [reply]
@ryanpeach 

The same problem here:

```
⬢ [Docker] ❯ nvim --headless "+MasonInstallAll" +qall
Error detected while processing command line:
E492: Not an editor command: MasonInstallAll
```

Have you managed to find a root cause of it?

**Comment by Eytan425** [comment]
Can you send the link to setup the neovim please

**Comment by qwertasyx** [comment]
I just noticed that as long as Mason is not loaded the MasonInstallAll cmd is there.
<img width="565" alt="image" src="https://github.com/user-attachments/assets/2c601d12-4ccb-484a-8a60-f0bece302de9" />

as soon as Mason gets loaded it disappears ... at least on one of my machines .. not sure what's going on here but maybe it helps someone who stumbles upon this thread in the future.

**Comment by Strongsloth** [comment]
the command MasonInstallAll is ONLY available in NvChad
